gobblin-data-management/src/main/java/org/apache/gobblin/data/management/retention/DatasetCleaner.java

```diff
@@ -155,6 +155,9 @@ public Void call() throws Exception {
         public void onFailure(Throwable throwable) {
           DatasetCleaner.this.finishCleanSignal.get().countDown();
           LOG.warn("Exception caught when cleaning " + dataset.datasetURN() + ".", throwable);
+          if (throwable instanceof OutOfMemoryError) {
```

Comment:
how come the job exits successfully when the throwable is added to `throwables` and a `RuntimeException` is thrown in `close`

Reply:
When OutofMemoryError occurs, there might be some heap space remaining for the execution of finally block. As we are consuming it in the throwable, the `RuntimeException` never gets executed in `close` resulting to go to finally block (where the Executor service gracefully shutdowns the process) 
Was able to replay this test case in my local.

Reply:
please add a comment of your finding

Reply:
Actual issue was due to the race condition with the countdownLatch, have fixed it. Added the RCA in the description